In [1]:
import numpy as np 
import pandas as pd
import os 
import sys 
import subprocess
if sys.platform == 'darwin':
    print("Current system is macOS")
    main_fold_path = '/Users/shanxiafeng/Documents/Project/Research/fnirs-prognosis/code/fnirs-treatment-response-prediction'
elif sys.platform == 'linux':
    print("Current system is Ubuntu")
    main_fold_path = '/home/jy/Documents/fnirs/treatment_response/fnirs-depression-deeplearning'
else:
    print("Current system is neither macOS nor Ubuntu")
os.chdir(main_fold_path)
from utils.fnirs_utils import print_md_table_val_test_AUC


Current system is Ubuntu


In [6]:


for itr in range(4, 14):
    bash_code = f"conda run -n tf python scripts/plot/DL/read_LOO_nestedCV_gnntr_auc.py --max 4 --K_FOLD 5 --result_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v{itr}l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6 "
    subprocess.run(bash_code, shell=True)



/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:13.869962: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:13.889102: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:13.889124: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:13.889135: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v4l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 2 ~ 4 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 66.0000  | 50.0000  | 82.0000  | 55.7143  | 83.1511  | 70.7813  | 95.5209  | 83.1509  |
| test_k_fold_0_AUC1_0.59_AUC2_0.59   | 58.5714  | 57.1429  | 60.0000  | 62.839

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:16.164988: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:16.184000: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:16.184019: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:16.184026: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v6l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 2 ~ 3 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 59.8571  | 35.7143  | 84.0000  | 62.0000  | 82.2535  | 69.3750  | 95.1320  | 82.2534  |
| test_k_fold_0_AUC1_0.63_AUC2_0.6   | 59.5714  | 57.1429  | 62.0000  | 64.2120

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:18.439731: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:18.458971: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:18.458992: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:18.458999: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v7l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 2 ~ 3 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 62.4286  | 42.8571  | 82.0000  | 60.2857  | 81.1650  | 67.6563  | 94.6737  | 81.1648  |
| test_k_fold_0_AUC1_0.6_AUC2_0.54   | 54.0000  | 50.0000  | 58.0000  | 59.9810

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:20.708248: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:20.727129: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:20.727147: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:20.727154: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v8l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 3 ~ 4 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 62.4286  | 42.8571  | 82.0000  | 60.2857  | 82.3924  | 69.2188  | 95.5660  | 82.3923  |
| test_k_fold_0_AUC1_0.61_AUC2_0.6   | 60.1429  | 64.2857  | 56.0000  | 61.4610

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:23.013252: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:23.032255: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:23.032275: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:23.032282: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v9l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 3 ~ 5 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 67.0000  | 50.0000  | 84.0000  | 65.0000  | 83.6597  | 71.5625  | 95.7570  | 83.6596  |
| test_k_fold_0_AUC1_0.62_AUC2_0.57   | 56.5714  | 57.1429  | 56.0000  | 60.038

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:25.282779: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:25.301694: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:25.301715: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:25.301722: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v10l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 7 ~ 8 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 64.4286  | 42.8571  | 86.0000  | 50.0000  | 84.5070  | 73.4375  | 95.5764  | 84.5068  |
| test_k_fold_0_AUC1_0.41_AUC2_0.42   | 42.2857  | 28.5714  | 56.0000  | 54.09

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:29.886020: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:29.905089: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:29.905115: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:29.905121: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v11l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 5 ~ 6 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 62.4286  | 42.8571  | 82.0000  | 60.8571  | 85.0226  | 74.6875  | 95.3577  | 85.0225  |
| test_k_fold_0_AUC1_0.47_AUC2_0.54   | 54.0000  | 50.0000  | 58.0000  | 59.98

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:32.171718: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:32.190652: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:32.190674: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:32.190681: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

val_fold_path results/gnn_transformer/prognosis_mix_hb/pretreatment_response/loocv_v12l1_rate_0.01_l2_rate_0.001_d_model_16_batch_size_64_n_layers_6/LOO_nested_CV/LOO_0
num_of_cv_folds 5
ALL_Y_pred_in_test ->  (320, 2)
the y_pred has been saved in allData/prognosis_mix_hb/pretreatment_response/MDDR
MMDT_score.shape (64,)
MAX_ITR: 4 ranging ( 5 ~ 6 )
Model name: gnn_transformer
specify_itr None
value_add_to_sensitivity_value 0.0
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| gnn_transformer   | 57.8571  | 35.7143  | 80.0000  | 56.4286  | 83.6563  | 71.4063  | 95.9063  | 83.6562  |
| test_k_fold_0_AUC1_0.59_AUC2_0.52   | 51.5714  | 57.1429  | 46.0000  | 52.63

/usr/bin/bash: /home/jy/miniconda3/envs/tf/lib/libtinfo.so.6: no version information available (required by /usr/bin/bash)
2024-04-16 14:49:34.466507: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-16 14:49:34.485778: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:7704] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-04-16 14:49:34.485799: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-04-16 14:49:34.485806: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1520] Unable to register cuBLAS factory: Attempting to register factory 

In [13]:
csv_path = 'results/gnn_transformer/prognosis_mix_hb/metrics.txt'

df = pd.read_csv(csv_path)
import pandas as pd

# Assuming 'df' is your DataFrame
# Calculating the balanced accuracy for each metric in the test set
balanced_test_accuracy = (df['Test Sensitivity'] + df['Test Specificity']) / 2

# Calculating the balanced accuracy for each metric in the validation set
balanced_validation_accuracy = (df['Validation Sensitivity'] + df['Validation Specificity']) / 2

df['Test Accuracy'] = balanced_test_accuracy
df['Validation Accuracy'] = balanced_validation_accuracy


# Calculating the mean and standard deviation for each metric in the test set
average_test_accuracy = df['Test Accuracy'].mean()
std_test_accuracy = df['Test Accuracy'].std()

average_test_sensitivity = df['Test Sensitivity'].mean()
std_test_sensitivity = df['Test Sensitivity'].std()

average_test_specificity = df['Test Specificity'].mean()
std_test_specificity = df['Test Specificity'].std()

average_test_auc = df['Test AUC'].mean()
std_test_auc = df['Test AUC'].std()

# Calculating the mean and standard deviation for each metric in the validation set
average_validation_accuracy = df['Validation Accuracy'].mean()
std_validation_accuracy = df['Validation Accuracy'].std()

average_validation_sensitivity = df['Validation Sensitivity'].mean()
std_validation_sensitivity = df['Validation Sensitivity'].std()

average_validation_specificity = df['Validation Specificity'].mean()
std_validation_specificity = df['Validation Specificity'].std()

average_validation_auc = df['Validation AUC'].mean()
std_validation_auc = df['Validation AUC'].std()

# Printing the results with mean and standard deviation
print("Average Test Metrics:")
print(f"Accuracy: {average_test_accuracy} (±{std_test_accuracy})")
print(f"Sensitivity: {average_test_sensitivity} (±{std_test_sensitivity})")
print(f"Specificity: {average_test_specificity} (±{std_test_specificity})")
print(f"AUC: {average_test_auc} (±{std_test_auc})")

print("\nAverage Validation Metrics:")
print(f"Accuracy: {average_validation_accuracy} (±{std_validation_accuracy})")
print(f"Sensitivity: {average_validation_sensitivity} (±{std_validation_sensitivity})")
print(f"Specificity: {average_validation_specificity} (±{std_validation_specificity})")
print(f"AUC: {average_validation_auc} (±{std_validation_auc})")

test_metrics = [average_test_accuracy, average_test_sensitivity, average_test_specificity, average_test_auc]
val_metrics = [average_validation_accuracy, average_validation_sensitivity, average_validation_specificity, average_validation_auc]

print_md_table_val_test_AUC('Mean', test_metrics, val_metrics)


Average Test Metrics:
Accuracy: 0.6328571428571428 (±0.03142135559299737)
Sensitivity: 0.43571428571428567 (±0.05270462766947301)
Specificity: 0.8299999999999998 (±0.019436506316150997)
AUC: 0.5992857142857143 (±0.04702606986731438)

Average Validation Metrics:
Accuracy: 0.8298838578125004 (±0.012694089454970334)
Sensitivity: 0.7053125000000005 (±0.023664410837326493)
Specificity: 0.9544552156250002 (±0.003440104679139837)
AUC: 0.829882626645267 (±0.01269407719856165)
| Model Name | Testing Set |             |             |             | Validation Set |             |             |             |
|------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
|            | Balanced Accuracy | Sensitivity | Specificity | AUC | Balanced Accuracy | Sensitivity | Specificity | AUC |
| Mean   | 63.2857  | 43.5714  | 83.0000  | 59.9286  | 82.9884  | 70.5313  | 95.4455  | 82.9883  |

| Metric         | Test Set Average | Validation Set Average |
|----------------|------------------|------------------------|
| Accuracy       | 0.7437 | 0.9101 |
| Sensitivity    | 0.4358 | 0.7063 |
| Specificity    | 0.8320 | 0.9552 |
| AUC            | 0.5993 | 0.8327 |
